### DISCUSSION TITLE
SSLCertVerificationError when adding HTTPS openai api url

### DISCUSSION THREAD

**Discussion by jtc1246** [open-webui/open-webui#3702] [General]
# Bug Report

## Description

**Bug Summary:**

ERROR:apps.openai.main:Connection error: Cannot connect to host api.openai.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')] When tring to add openai api key and url.

But it has the ability to judge whether an openai api key is correct. (If I input a correct key, it will show Server connection verified, if input an incorrect key, it will show OpenAI: Network Problem) So this is very strange.

And api url with http is always ok.

**Steps to Reproduce:**
Go to Admin Panel / Settings / Connection, set api key and url. (my url is just https://api.openai.com/v1)

**Expected Behavior:**
Can add openai's models.

**Actual Behavior:**
ERROR:apps.openai.main:Connection error: Cannot connect to host api.openai.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')]
And no openai models shown.

## Environment

- **Open WebUI Version:** v0.3.7
- **Ollama (if applicable):** 0.1.48

- **Operating System:** macos 14.1.1 (23B81)
- **Browser (if applicable):** chrome 125.0.6422.142 (Official Build) (arm64)

## Reproduction Details

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [x] I have included the Docker container logs.

## Logs and Screenshots

**Browser Console Logs:**
<img width="1020" alt="Screenshot 2024-07-08 at 02 43 08" src="https://github.com/open-webui/open-webui/assets/43341500/2c6040cc-7cf1-49a6-a70d-470c4dab02de">

**Docker Container Logs:**

```
INFO:     127.0.0.1:63732 - "GET /openai/models/0 HTTP/1.1" 200 OK
INFO:apps.openai.main:get_all_models()
ERROR:apps.openai.main:Connection error: Cannot connect to host api.openai.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')]
INFO:apps.ollama.main:get_all_models()
INFO:     127.0.0.1:63732 - "GET /api/models HTTP/1.1" 200 OK
INFO:apps.openai.main:get_all_models()
ERROR:apps.openai.main:Connection error: Cannot connect to host api.openai.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')]
INFO:config:Saving 'OPENAI_API_BASE_URLS' to config.json
INFO:     127.0.0.1:63959 - "POST /openai/urls/update HTTP/1.1" 200 OK
INFO:config:Saving 'OPENAI_API_KEYS' to config.json
INFO:     127.0.0.1:63959 - "POST /openai/keys/update HTTP/1.1" 200 OK
ERROR:apps.openai.main:401 Client Error: Unauthorized for url: https://api.openai.com/v1/models
Traceback (most recent call last):  // here is just input the incorrect openai api key
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/apps/openai/main.py", line 324, in get_models
    r.raise_for_status()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://api.openai.com/v1/models
INFO:     127.0.0.1:63959 - "GET /openai/models/0 HTTP/1.1" 500 Internal Server Error
INFO:apps.openai.main:get_all_models()
ERROR:apps.openai.main:Connection error: Cannot connect to host api.openai.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')]
INFO:apps.ollama.main:get_all_models()
INFO:     127.0.0.1:63959 - "GET /api/models HTTP/1.1" 200 OK
```

**Screenshots (if applicable):**
[Attach any relevant screenshots to help illustrate the issue]

## Installation Method

pip

## Additional Information

I have tried using requests in this python, no ssl certificate problem.

```
>>> response = requests.get("https://www.google.com")
>>> print(response.status_code)
200
>>> response = requests.get("https://api.openai.com")
>>> print(response.status_code)
421
```

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by jtc1246** [comment]
Solved by following:

In site-packages/apps/openai/main.py, change 2 functions.

In async def fetch_url(url, key):, line about 191, change 
```
        async with aiohttp.ClientSession(timeout=timeout, trust_env=True) as session:
```
to
```
        connector = aiohttp.TCPConnector(ssl=False)
        async with aiohttp.ClientSession(timeout=timeout, trust_env=True, connector=connector) as session:
```

In async def generate_chat_completion(, line about 466, change
```
try:
        session = aiohttp.ClientSession(trust_env=True)
```
to
```
try:
        connector = aiohttp.TCPConnector(ssl=False)
        session = aiohttp.ClientSession(trust_env=True, connector=connector)
```

**Comment by justinh-rahb** [comment]
Ya that's not going to be patched in, you've disabled SSL verification. There's no bug here, if this was an issue it'd be immediately noticed by every developer before it ever hit `main` branch. I'd instead wonder why your SSL is failing to verify.. maybe your traffic is being MITM'd? 🤷‍♂️

**Comment by jtc1246** [comment]
It seems that verification and actual usage later use different http libraries, one is requests, one is aiohttp. They may have different set of root certificates. Aiohttp may not recognize the certificate of openai.

**Comment by walloutlet** [comment]
Instead of a patch, could we ask for a feature to add a checkbox to disable verification? Locally hosted instances of LLM's that use the OpenAI API endpoints to access sometimes have self-signed certificates.

**Reply by Gabgobie** [reply]
TL;DR: I'd really like to see an env var to disable validation and if needed an env var to add trusted CAs.

I was already looking through the docs for an env var to disable validation. In my dev env I am currently working with self signed certs. I will replace them with ones signed by my own CA but that is part of the roadmap for now. Even then I am not entirely sure if the issue will be fixed as I think to remember python using its own trust store instead of the system/container store. There was also no env var to add CA trust.

This is currently keeping me from protecting the connection between ollama and OI as well as implementing OIDC. I do think that an encrypted connection that is not validating the certificate is better than an unencrypted connection. OFC validation is preferable but that would require me to be able to add my CA to the trust store which I have yet to validate to be possible.

Edit:
Reading through #1398 it seems to be possible to add CA trust, although the process is currently a little clumsy as you have to do so for the requests and the httpx library in separate variables. While not very pretty, it is presumably in working condition and therefore not really a priority. It would be nice to have an env var to disable validation, which is currently missing from what I can tell so one doesn't have to spin up a full PKI to do some testing etc. I guess for me this is as good a time as any to finally spin up my PKI so all of my projects will be on ice until that is done and I'll report back with my findings.

**Comment by tanu360** [comment]
New Location: site-packages/open_webui/routers/openai.py

**Comment by you-can-change** [comment]
I am from china, can not visit openai, 

I use `pip install open-webui`, 

```bash
# pip list | grep open-webui
open-webui                               0.5.10
```

When I logged in, the web page remained blank and I received the following logs:
```
ERROR [open_webui.routers.openai] Connection error: Cannot connect to host api.openai.com:443 ssl:default [None]
INFO  [open_webui.routers.ollama] get_all_models()
```

![image](https://github.com/user-attachments/assets/ce2ad9d1-7c35-48db-9115-bbb5aab9c892)

modify `/mnt/workspace/miniconda3/envs/open_web-ui/lib/python3.11/site-packages/open_webui/utils/models.py`

```python
async def get_all_base_models(request: Request):
    function_models = []
    openai_models = []
    ollama_models = []
    # comment these lines
    #if request.app.state.config.ENABLE_OPENAI_API:
    #    openai_models = await openai.get_all_models(request)
    #    openai_models = openai_models["data"]
```

then restart open-webui, it's work fine.


Modifying the `ENABLE_OPENAI_API` parameter may also work, but I searched and couldn't find it. I tried `exporting  ENABLE_OPENAI_API=false` before starting, but it didn't take effect. Let's do this for now

**Comment by richtong** [comment]
HOw was this resolved, we have the following issues:

1. Access OpenAI in Docker works
2. Access with Curl works
3. Access with pipx installed open-webui fails with the error above
4. We have a vanilla pipx install open-webui on 0.5.20 and this still fails. Sigh

Note that this is happening on 1 of 4 machines, so it sure seems like there is some sort of cache or something that is not working. These machines are in four different geographies, we don't think that it is related to that, but thought I would add.

@pauldavis and @richtong are working on this, any help appreciated. Scratching our heads on this one.

**Comment by lucashmorais** [comment]
A solution could be to force every call to `aiohttp.ClientSession()` in `openai.py` to use a certificate bundle from `certifi`:

```python
import ssl
import certifi

# Function added to `openai.py` to get the `certifi` bundle
def get_good_connector():
    ca_bundle_path = certifi.where()

    # Create a custom SSLContext
    ssl_context = ssl.create_default_context()
    ssl_context.load_verify_locations(cafile=ca_bundle_path)

    # Configure a TCPConnector with the custom SSLContext
    connector = aiohttp.TCPConnector(ssl=ssl_context)
    return connector
```

Then, add `connector=get_good_connector()` to every `aiohttp.ClientSession` of the same file, like in what follows:

```python
        async with aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=AIOHTTP_CLIENT_TIMEOUT_MODEL_LIST),
            connector=get_good_connector()
        ) as session:
```

Make sure that your certificate bundle is up-to-date with `pip install --upgrade certifi`, called from the same environment running the open-webui server.

**Comment by kbaegis** [comment]
https://github.com/certifi/python-certifi?tab=readme-ov-file#additionremoval-of-certificates

Certifi doesn't support new CAs. REQUESTS_CA_BUNDLE doesn't appear to work as an override, and running update-ca-certificates doesn't appear to carry over to the backend.

The documentation at https://docs.openwebui.com/tutorials/integrations/custom-ca appears to be defunct.

**Comment by kbaegis** [comment]
~PR submitted. https://github.com/open-webui/open-webui/pull/15868~
